firedrake/cofunction.py

```diff
@@ -189,15 +191,22 @@ def assign(self, expr, subset=None):
         """
         expr = ufl.as_ufl(expr)
         if isinstance(expr, ufl.classes.Zero):
-            self.dat.zero(subset=subset)
+            with stop_annotating(modifies=(self,)):
+                self.dat.zero(subset=subset)
             return self
         elif (isinstance(expr, Cofunction)
               and expr.function_space() == self.function_space()):
+            # do not annotate in case of self assignment
+            annotate = annotate_tape() and self != expr
+            if annotate:
+                self.block_variable = self.create_block_variable()
+                self.block_variable._checkpoint = DelegatedFunctionCheckpoint(expr.block_variable)
```

Comment:
The approach of using `self.block_variable = expr.block_variable` is not effective for general cases. Particularly, for variational solvers. `self.block_variable = expr.block_variable` results in the loss of the `self.block_variable.output` identity, which must match the identity of `self`. Modifying either `expr.count` or `self.count` is not safe. Therefore, the current code proposes to create a block variable and use `DelegatedFunctionCheckpoint` to avoid saving the assigned data twice.